mtags/src/main/scala-2/scala/meta/internal/pc/ExtractMethodProvider.scala

```diff
@@ -0,0 +1,182 @@
+      renames = re
+    )
+    def prettyType(tpe: Type) =
+      metalsToLongString(tpe.widen.finalResultType, history)
+
+    def pathTo(tree: Tree, range: l.Range): List[Tree] = {
+      def loop(tree: Tree): List[Tree] = {
+        tree.children
+          .filter(_.pos.isDefined)
+          .find(t => t.pos.toLSP.encloses(range)) match {
+          case Some(t) =>
+            t :: loop(t)
+          case None => Nil
+        }
+      }
+      loop(tree).reverse
+    }
+    def extractFromBlock(t: Tree): List[Tree] =
+      t match {
+        case Block(stats, expr) =>
+          (stats :+ expr).filter(stat => range.encloses(stat.pos.toLSP))
+        case temp: Template =>
+          temp.body.filter(stat => range.encloses(stat.pos.toLSP))
+        case other => List(other)
+      }
+
+    def valsOnPath(ts: List[Tree]): List[(TermName, String)] = {
+      ts.flatMap(t =>
+        t match {
+          case Block(stats, expr) => valsOnPath(stats :+ expr)
+          case Template(_, _, body) => valsOnPath(body)
+          case ValDef(_, name, tpt, _) if tpt.tpe != null =>
+            List((name, prettyType(tpt.tpe)))
+          case _ => Nil
+        }
+      )
+    }
+
+    def genName(path: List[Tree]): String = {
+      def defsOnPath(ts: List[Tree]): Set[String] = {
+        ts.flatMap(t =>
+          t match {
+            case Block(stats, expr) => defsOnPath(stats :+ expr)
+            case Template(_, _, body) => defsOnPath(body)
+            case DefDef(_, name, _, _, _, _) =>
+              Seq(name.toString)
+            case _ => Nil
+          }
+        ).toSet
+      }
+      val usedNames = defsOnPath(path)
+      if (!usedNames("newMethod")) "newMethod"
+      else {
+        var i = 0
+        while (usedNames(s"newMethod$i")) {
+          i += 1
+        }
+        s"newMethod$i"
+      }
+    }
+
+    def localRefs(ts: List[Tree]): Set[TermName] = {
+      val names = Set.newBuilder[TermName]
+      val defns = mutable.Set[TermName]()
+
+      def traverse(tree: Tree): Unit = tree match {
+        case Ident(name) =>
+          if (!defns(name.toTermName)) names += name.toTermName
+        case Select(qualifier, name) =>
+          if (!defns(name.toTermName)) names += name.toTermName
+          traverse(qualifier)
+        case ValDef(_, name, _, rhs) =>
+          defns.add(name)
+          traverse(rhs)
+        case _ => tree.children.foreach(traverse(_))
+      }
+
+      ts.map(traverse(_))
+
+      names.result()
+    }
+
+    def adjustIndent(
+        line: String,
+        newIndent: String,
+        oldIndent: Int
+    ): String = {
+      var i = 0
+      val additional = if (newIndent.indexOf("\t") != -1) "\t" else "  "
+      while ((line(i) == ' ' || line(i) == '\t') && i < oldIndent) {
+        i += 1
+      }
+      newIndent + additional + line.drop(i)
+    }
+
+    val path = pathTo(unit.body, range)
+    val edits =
+      for {
+        enclosing <- path.find(_.pos.toLSP.encloses(range))
```

Comment:
```suggestion
        enclosing <- path.headOption
```
since we already check for enclosing?